repair/repair.cc

```diff
@@ -2104,6 +2101,32 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
             std::vector<gms::inet_address> nodes;
             auto master_shard_id = m.master_shard_id;
             auto range = m.range;
+            bool select = true;
+            if (!ranges_specified.empty()) {
+                select = false;
+                for (auto& r : ranges_specified) {
+                    // When the management tool, e.g., scylla manager, uses
+                    // ranges option to select which ranges to repair for a
+                    // tablet table to schedule repair work, it needs to
+                    // disable tablet migration to make sure the node and token
+                    // range mapping does not change. Here, we check if the
+                    // user specified range and the range owned by the node
+                    // matches exactly, ignoring the case the two ranges
+                    // match partially due to tablet migration.
+                    if (range == r) {
```

Comment:
We need to intersect each tablet's range with the specified ranges to find what we should repair, not assume that they align. Consider merge and split.

Reply:
Or someone repairing a single token to fix a problem they see.

Reply:
I thought about intersection too but I think it is not needed. The management tool is supposed to send the ranges matches exactly. In case the token ranges requested are different than the actual ranges the node owns, it is an indication that the token range mapping for the tablet has changed. It could be multiple reasons for such change, like the one you mentioned merge and split. It could also be adding or removing of nodes.  In case of token ownership change, the repair scheduling based on the old mapping is not useful anymore. Doing the intersection and fix part of the requested ranges is not going to help much, so I think it is better to skip current wrong scheduling. We can reply on the next scheduling of repair.

Reply:
The management tool doesn't know the tablet ranges, because they are split and merged dynamically. It's also better to avoid polluting the logic here with special cases and instead have one case.

Reply:
There are two approaches to schedule repair cluster wide. 1) What we do today, the management tool gets a token range ownership mapping, it decides who is going to repair which token range, avoiding same range to be repaired more than once and placing repair jobs in a smart way not to overwhelm nodes. We have 04773bd1dfe3218c236505e7e416d65a738c157f to allow the exiting apis to work with tablet. 2) Schedule tablet repair inside scylla core which is not part of 6.0 release. 

This patch is for the first approach which is needed for 6.0 release. Intersecting is introducing a special cases. What we do in this patch is actually making the logic much simpler and easier to reason about than intersecting ranges.

Reply:
it looks like this loop could be simplified by using a predicate like
`select = should_select_range(ranges_specified, range)`
If `ranges_specified` would be turned into an unordered_set it would even be easier to check the range's existence in it, since you're looking for an exact match, not for an intersection.

Reply:
we discussed that in today's coordination meeting.
- there is no expectation that the user pauses tablet migration during repair.
- that should be tolerated since we need to enforce tombstone_gc_mode = repair (#16627)
- given that tablets might split (and merge in the future) during repair we should also support intersection of tablet ranges, not only exact match.

Reply:
Say the range [1, 100] is owned by a node n1 when the scheduling runs, before n1 has the chance to run repair for range [1, 100], the split happens, so n1 owns only [1, 50]. The intersection is only helping part of the range while the other part is completely ignored. The point is that, the intersection is not solving the problem and it only makes it a little bit better. It is similar for the merge as well.  At the end of the day, we still need the next scheduling of repair to complete the whole ranges, no matter if it is split,  merge or migration. 

When the next repair is scheduled, the range [1, 50] is repaired again along with the unrepaired range [50, 100]. This makes range [1, 50] repaired twice while range [50, 100] is repaired once, which is not a correctness issue. But if we do not do the intersection, we could finish the wrong scheduling faster because the wrong scheduling is ignored. In the end, we finish the whole range repair faster.  

Split and merge are not supposed to happen DURING repair, we hold the erm at the beginning of repair. Doesn't it prevent the tablet split/merge/migration?

Reply:
> When the next repair is scheduled, the range [1, 50] is repaired again along with the unrepaired range [50, 100]. This makes range [1, 50] repaired twice while range [50, 100] is repaired once, which is not a correctness issue. But if we do not do the intersection, we could finish the wrong scheduling faster because the wrong scheduling is ignored. In the end, we finish the whole range repair faster.

I think we should still do our best to repair as much as possible. There is no guarantee that the next repair won't hit the same race. The fact that concurrently migrating tablets may be skipped is already bad. Although we say we can live with that, we shouldn't use it as an excuse for introducing more cases like that.

There is also the case when you repair without -pr, in which case repair may be expected to always repair the tablet. If we don't intersect, this expectation will not be true. Avi brought up the use case where requested range may not align with tablet range (e.g. single token).

> Split and merge are not supposed to happen DURING repair, we hold the erm at the beginning of repair. Doesn't it prevent the tablet split/merge/migration?

It prevents migration during repair, but ranges are determined by the request-maker before that.

Reply:
> I think we should still do our best to repair as much as possible. There is no guarantee that the next repair won't hit the same race. The fact that concurrently migrating tablets may be skipped is already bad. Although we say we can live with that, we shouldn't use it as an excuse for introducing more cases like that.
> 
> There is also the case when you repair without -pr, in which case repair may be expected to always repair the tablet. If we don't intersect, this expectation will not be true. 

Without -pr nor -ranges options, repair will repair all the tablets on the given host. There is no need to intersect at all.

> Avi brought up the use case where requested range may not align with tablet range (e.g. single token).

User is not supposed to send a random token range to a host. 

But, anyway, since the intersection is not making things worse, I think we can live with intersection for now.

Reply:
> > I think we should still do our best to repair as much as possible. There is no guarantee that the next repair won't hit the same race. The fact that concurrently migrating tablets may be skipped is already bad. Although we say we can live with that, we shouldn't use it as an excuse for introducing more cases like that.
> > There is also the case when you repair without -pr, in which case repair may be expected to always repair the tablet. If we don't intersect, this expectation will not be true.
> 
> Without -pr nor -ranges options, repair will repair all the tablets on the given host. There is no need to intersect at all.

I meant without -pr but with -ranges.

> 
> > Avi brought up the use case where requested range may not align with tablet range (e.g. single token).
> 
> User is not supposed to send a random token range to a host.

Do we mention that in the documentation or reject such a request?

> 
> But, anyway, since the intersection is not making things worse, I think we can live with intersection for now.

Reply:
We still allow a random range to be passed to a repair request.  I had a PR in the past to reject such requests, https://github.com/scylladb/scylladb/pull/8142.  People had different opinions on this. It was not merged. 

With tablet, the wrong range could be passed not due to a admin error but migration, so we'd better continue to allow it.